File: Polars.FSharp/Expr/CatOps.fs
namespace Polars.FSharp

open Polars.NET.Core

type [<Struct>] CategoricalOps(handle: ExprHandle) =
    /// <summary>
    /// Get the categories stored in this data type.
    /// </summary>
    member _.GetCategories() = new Expr(PolarsWrapper.CatGetCategories handle)
    /// <summary>
    /// Return the byte-length of the string representation of each value.
    /// </summary>
    member _.LenBytes() = new Expr(PolarsWrapper.CatLenBytes handle)
    /// <summary>
    /// Return the number of characters of the string representation of each value.
    /// </summary>
    member _.LenChars() = new Expr(PolarsWrapper.CatLenChars handle)
    /// <summary>
    /// Check if string representations of values start with a substring.
    /// </summary>
    member _.StartsWith prefix = new Expr(PolarsWrapper.CatStartsWith(handle,prefix))
    /// <summary>
    /// Check if string representations of values end with a substring.
    /// </summary>
    member _.EndsWith suffix = new Expr(PolarsWrapper.CatEndsWith(handle,suffix))
    /// <summary>
    /// Get the physical values of a categorical or enum data type.
    /// </summary>
    member _.Physical() = new Expr(PolarsWrapper.CatPhysical handle)
    /// <summary>
    /// Convert to a categorical or enum dtype.
    /// The input must be of the physical type of the categorical or enum dtype.
    /// </summary>
    /// <param name="dtype">The target categorical or enum dtype.</param>
    /// <param name="strict">Whether to panic when encountering an illegal category.</param>
    member _.To(dtype:DataTypeExpr,?strict) =
        let st = defaultArg strict true 
        new Expr(PolarsWrapper.CatTo(handle,dtype.CloneHandle(),st))

File: Polars.CSharp/Series/CatOps.cs
using Polars.NET.Core;
using Pl = Polars.CSharp.Polars;

namespace Polars.CSharp;

/// <summary>
/// Operations on Categorical series. Access via <see cref="Series.Cat"/>.
/// </summary>
public readonly struct SeriesCategoricalOps
{
    private readonly Series _series;
    internal SeriesCategoricalOps(Series series) { _series = series; }
    private Series Apply(Func<Expr, Expr> op) 
        => _series.ApplyExpr(op(Pl.Col(_series.Name)));
    /// <inheritdoc cref="CategoricalOps.GetCategories"/>
    public Series GetCategories() => Apply(e=>e.Cat.GetCategories());
    /// <inheritdoc cref="CategoricalOps.LenBytes"/>
    public Series LenBytes() => Apply(e=>e.Cat.LenBytes());
    /// <inheritdoc cref="CategoricalOps.LenChars"/>
    public Series LenChars() => Apply(e=>e.Cat.LenChars());
    /// <inheritdoc cref="CategoricalOps.StartsWith"/>
    public Series StartsWith(string prefix) => Apply(e=>e.Cat.StartsWith(prefix));
    /// <inheritdoc cref="CategoricalOps.EndsWith"/>
    public Series EndsWith(string suffix) => Apply(e=>e.Cat.EndsWith(suffix));
    /// <inheritdoc cref="CategoricalOps.Physical"/>
    public Series Physical() => Apply(e=>e.Cat.Physical());
    /// <inheritdoc cref="CategoricalOps.To"/>
    public Series To(IntoDataTypeExpr dtype,bool strict=true) => Apply(e=>e.Cat.To(dtype,strict)); 
}